java/ql/test/query-tests/security/CWE-502/DelayedDeserialization.java

```diff
@@ -0,0 +1,30 @@
+import java.io.ByteArrayInputStream;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.ObjectInputStream;
+import java.net.Socket;
+
+public class DelayedDeserialization {
+
+    public void test(Socket socket) throws Exception {
+        DataInputStream in = new DataInputStream(socket.getInputStream());
+        ObjectHolder holder = new ObjectHolder();
+        holder.init(in);
+        holder.getObject();
+    }
+}
+
+class ObjectHolder {
+
+    private final byte[] buffer = new byte[1024];
+
+    public void init(DataInput input) throws Exception {
+        input.readFully(buffer, 0, buffer.length);
+    }
+
+    public Object getObject() throws Exception {
+        ByteArrayInputStream bais = new ByteArrayInputStream(buffer);
+        ObjectInputStream ois = new ObjectInputStream(bais);
+        return ois.readObject(); // $unsafeDeserialization
+    }
+}
```

Comment:
Maybe it would be good to also add a test where an assignment to a `byte[]` field transfers taint, e.g.:
(assuming that is the reason why you chose `FieldAccess` instead of the more specific `FieldRead` for your `byteArrayFieldFlowStep` predicate above)
```java
...

class ObjectHolder {
    private byte[] buffer;

    public void init(byte[] tainted) throws Exception {
        buffer = tainted;
    }

    public Object getObject() throws Exception {
        ByteArrayInputStream bais = new ByteArrayInputStream(buffer);
        ObjectInputStream ois = new ObjectInputStream(bais);
        return ois.readObject(); // $unsafeDeserialization
    }
}
```

Reply:
Thanks for the idea!